### PR TITLE
Support for Structured Authorization for `Shoot`s >= Kubernetes v1.30

### LINKED ISSUE #9740: Support for Structured Authorization for `Shoot`s >= Kubernetes v1.30
**How to categorize this issue?**
 /area control-plane
/kind api-change
/kind enhancement

**What would you like to be added**:

Hello Team 👋 

We would like to benefit from the ability to use custom [Kubernetes authorization webhooks](https://kubernetes.io/docs/reference/access-authn-authz/webhook/) in order to give a gardener provided control plane the ability to defer Authorization decisions to an external authorization engine like OpenFGA.

The ability to be able to configure the necessary kubernetes API server args via the Shoot CRD would be needed in order to achieve our desired result

**Why is this needed**:

 This change is needed to enable users of gardener to profit form all the features that kubernetes provides in regards to authorization modes. We would like to use this feature within the context of the ORA project, so all contributers are able to use a central authorization engine, which is able to give authorization decisions in every layer of the product.

Thanks a lot and let me know in case any other input is needed!

### PR DESCRIPTION
**Pull request by rfranzke**
**How to categorize this PR?**

/area security
/kind enhancement

**What this PR does / why we need it**:
Similar to #10244, this PR adds support for structured authorization for shoots >= Kubernetes 1.30.
Read https://github.com/rfranzke/gardener/blob/structured-authz/docs/usage/shoot/shoot_access.md#structured-authorization to learn what has been implemented in this PR.

**Which issue(s) this PR fixes**:
Fixes https://github.com/gardener/gardener/issues/9740

**Release note**:

```feature user
Structured authorization configuration can now be set by creating a `ConfigMap` with the `AuthorizationConfiguration` file set in the `config.yaml` data key and referencing it (in the `Shoot` via `.spec.kubernetes.kubeAPIServer.structuredAuthorization`, in the `Garden` via `.spec.virtualCluster.kubernetes.kubeAPIServer.structuredAuthorization` for Kubernetes versions `>= v1.30`. Read all about it [here](https://github.com/gardener/gardener/tree/master/docs/usage/shoot/shoot_access.md#structured-authorization).
```

### CODE REVIEW

File: docs/usage/shoot/shoot_access.md [hunk 1]
```diff
@@ -182,7 +182,67 @@ data:
...
+```
+
+The user is responsible for the validity of the configured authorizers.
+
+> [!NOTE]
+> You can have one or more authorizers of type `Webhook` (no other types are supported).
+>
+> You are not allowed to specify the `authorizers[].webhook.connectionInfo` field.
+> Instead, provide a kubeconfig file containing the server address (and optionally, credentials that can be used by `kube-apiserver` in order to authenticate with the webhook server) by creating a `Secret` containing the kubeconfig (in the `.data.kubeconfig` key).
+> Reference this `Secret` by adding it `.spec.kubernetes.kubeAPIServer.structuredAuthorization.kubeconfigs[]` (choose the proper `authorizerName`, see example above).
```
Comment (timuthy) [inline docs/usage/shoot/shoot_access.md]:
```suggestion
> Reference this `Secret` by adding it to `.spec.kubernetes.kubeAPIServer.structuredAuthorization.kubeconfigs[]` (choose the proper `authorizerName`, see example above).
```

File: pkg/apis/core/types_shoot.go [hunk 2]
```diff
@@ -636,6 +634,10 @@ type KubeAPIServerConfig struct {
 	DefaultUnreachableTolerationSeconds *int64
 	// EncryptionConfig contains customizable encryption configuration of the API server.
 	EncryptionConfig *EncryptionConfig
+	// StructuredAuthentication contains configuration settings for structured authentication to the kube-apiserver.
+	StructuredAuthentication *StructuredAuthentication
+	// StructuredAuthorization contains configuration settings for structured authorization to the kube-apiserver.
```
Comment (timuthy) [inline pkg/apis/core/types_shoot.go]:
```suggestion
	// StructuredAuthorization contains configuration settings for structured authorization for the kube-apiserver.
```

Nit: Sounds more natural FMPOV. Also the struct doc string says "for".

File: pkg/apis/core/validation/shoot_test.go [hunk 3]
```diff
@@ -3136,6 +3136,63 @@ var _ = Describe("Shoot Validation Tests", func() {
...
+				shoot.Spec.Kubernetes.KubeAPIServer.FeatureGates = map[string]bool{
+					"StructuredAuthorizationConfiguration": false,
+				}
+
+				Expect(ValidateShoot(shoot)).To(ConsistOf(PointTo(MatchFields(IgnoreExtras, Fields{
+					"Type":   Equal(field.ErrorTypeForbidden),
+					"Field":  Equal("spec.kubernetes.kubeAPIServer.structuredAuthorization"),
+					"Detail": Equal("requires feature gate StructuredAuthorizationConfiguration to be enabled"),
+				}))))
+			})
```
Comment (timuthy) [inline pkg/apis/core/validation/shoot_test.go:3203]: Only failed scenarios are tested here, create an success case as well?

File: pkg/admissioncontroller/webhook/auth/seed/graph/eventhandler_shoot.go [hunk 4]
```diff
@@ -145,6 +146,18 @@ func (g *graph) handleShootCreateOrUpdate(ctx context.Context, shoot *gardencore
 			configMapVertex := g.getOrCreateVertex(VertexTypeConfigMap, shoot.Namespace, shoot.Spec.Kubernetes.KubeAPIServer.StructuredAuthentication.ConfigMapName)
 			g.addEdge(configMapVertex, shootVertex)
 		}
+
+		if len(v1beta1helper.GetShootAuthorizationConfigurationConfigMapName(kubeAPIServer)) > 0 {
+			configMapVertex := g.getOrCreateVertex(VertexTypeConfigMap, shoot.Namespace, shoot.Spec.Kubernetes.KubeAPIServer.StructuredAuthorization.ConfigMapName)
```
Comment (timuthy) [inline pkg/admissioncontroller/webhook/auth/seed/graph/eventhandler_shoot.go]:
```suggestion
if configMapName := v1beta1helper.GetShootAuthorizationConfigurationConfigMapName(kubeAPIServer); len(configMapName) > 0 {
			configMapVertex := g.getOrCreateVertex(VertexTypeConfigMap, shoot.Namespace, configMapName)
```

Nit: Slightly more stable to use the helpers output.

File: pkg/admissioncontroller/webhook/admission/authorizationconfig/handler.go [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+}
+
+func (h *Handler) admitShoot(ctx context.Context, request admission.Request) admission.Response {
+	shoot := &gardencore.Shoot{}
+	if err := runtime.DecodeInto(internalDecoder, request.Object.Raw, shoot); err != nil {
+		return admission.Errored(http.StatusInternalServerError, err)
+	}
+
+	if shoot.DeletionTimestamp != nil {
+		// don't mutate shoot if it's already marked for deletion, otherwise gardener-apiserver will deny the user's/
```
Comment (timuthy) [inline pkg/admissioncontroller/webhook/admission/authorizationconfig/handler.go]: What's mutated in that webhook?

File: pkg/admissioncontroller/webhook/admission/authorizationconfig/handler.go [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+	return admissionwebhook.Allowed("resource is not *core.gardener.cloud/v1beta1.Shoot or *v1.ConfigMap")
+}
+
+func (h *Handler) admitShoot(ctx context.Context, request admission.Request) admission.Response {
+	shoot := &gardencore.Shoot{}
+	if err := runtime.DecodeInto(internalDecoder, request.Object.Raw, shoot); err != nil {
+		return admission.Errored(http.StatusInternalServerError, err)
+	}
+
+	if shoot.DeletionTimestamp != nil {
```
Comment (timuthy) [inline pkg/admissioncontroller/webhook/admission/authorizationconfig/handler.go:97]: Is there a real reason not to validate shoots in deletion? One could still break the API server deployment with misconfiguration during deletion.

File: pkg/component/kubernetes/apiserver/authorization.go [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+
+	volumeMountPathStructuredAuthorizationConfig             = "/etc/kubernetes/structured/authorization"
+	volumeMountPathStructuredAuthorizationWebhookKubeconfigs = "/etc/kubernetes/structured/authorization-kubeconfigs" // #nosec G101 -- No credential.
+
+	configMapAuthorizationConfigDataKey = "config.yaml"
+)
+
+func (k *kubeAPIServer) reconcileConfigMapAuthorizationConfig(ctx context.Context, configMap *corev1.ConfigMap) error {
+	if value, ok := k.values.FeatureGates["StructuredAuthorizationConfiguration"]; (ok && !value) ||
+		versionutils.ConstraintK8sLess130.Check(k.values.Version) {
```
Comment (timuthy) [inline pkg/component/kubernetes/apiserver/authorization.go]: Nit: Extract to function.

File: pkg/component/kubernetes/apiserver/authorization.go [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+	}
+
+	configMap.Data = map[string]string{configMapAuthorizationConfigDataKey: string(data)}
+	utilruntime.Must(kubernetesutils.MakeUnique(configMap))
+	return client.IgnoreAlreadyExists(k.client.Client().Create(ctx, configMap))
+}
+
+func (k *kubeAPIServer) handleAuthorizationSettings(deployment *appsv1.Deployment, configMapAuthorizationConfig *corev1.ConfigMap, secretWebhooksKubeconfigs *corev1.Secret) {
+	if value, ok := k.values.FeatureGates["StructuredAuthorizationConfiguration"]; (!ok || value) &&
+		!versionutils.ConstraintK8sLess130.Check(k.values.Version) {
```
Comment (timuthy) [inline pkg/component/kubernetes/apiserver/authorization.go]: Nit: Switch if-else blocks, so it'll be similar to the condition in `reconcileConfigMapAuthorizationConfig`.

File: pkg/component/kubernetes/apiserver/authorization.go [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+		versionutils.ConstraintK8sLess130.Check(k.values.Version) {
+		return nil
+	}
+
+	authorizationConfiguration := &apiserverv1beta1.AuthorizationConfiguration{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: apiserverv1beta1.ConfigSchemeGroupVersion.String(),
+			Kind:       "AuthorizationConfiguration",
+		},
+		Authorizers: []apiserverv1beta1.AuthorizerConfiguration{{Type: "RBAC", Name: "rbac"}},
```
Comment (timuthy) [inline pkg/component/kubernetes/apiserver/authorization.go:51]: I think it would be beneficial to document that RBAC and Node are still configured behind the scenes.

File: pkg/admissioncontroller/webhook/admission/authorizationconfig/handler.go [hunk 10]
```diff
@@ -0,0 +1,250 @@
...
+			authConfig.Authorizers[i].Webhook.ConnectionInfo.Type = apiserver.AuthorizationWebhookConnectionInfoTypeInCluster
+		}
+	}
+
+	if verifyKubeconfigReferences {
+		for _, authorizerName := range authorizerNames {
+			if !slices.ContainsFunc(kubeconfigReferences, func(ref gardencore.AuthorizerKubeconfigReference) bool {
+				return ref.AuthorizerName == authorizerName
+			}) {
+				return http.StatusUnprocessableEntity, fmt.Errorf("provided invalid authorization configuration: must provide kubeconfig secret name reference for webhook authorizer %q", authorizerName)
```
Comment (timuthy) [inline pkg/admissioncontroller/webhook/admission/authorizationconfig/handler.go:229]: Where is the kubeconfig secret checked, i.e. the secret exists + `data["kubeconfig"]` is available?

File: pkg/apis/core/v1beta1/types_shoot.go [hunk 11]
```diff
@@ -923,11 +927,28 @@ type AuditPolicy struct {
...
 	ConfigMapName string `json:"configMapName" protobuf:"bytes,1,opt,name=configMapName"`
 }
 
+// StructuredAuthorization contains authorization config for kube-apiserver.
+type StructuredAuthorization struct {
+	// ConfigMapName is the name of the ConfigMap in the project namespace which contains AuthorizationConfiguration for
+	// the kube-apiserver.
+	ConfigMapName string `json:"configMapName" protobuf:"bytes,1,opt,name=configMapName"`
+	// Kubeconfigs is a list of references for kubeconfigs for the authorization webhooks.
+	Kubeconfigs []AuthorizerKubeconfigReference `json:"kubeconfigs" protobuf:"bytes,2,rep,name=kubeconfigs"`
```
Comment (timuthy) [inline pkg/apis/core/v1beta1/types_shoot.go:941]: Don't we need to adjust the [reference controller](https://github.com/gardener/gardener/tree/master/pkg/controllermanager/controller/shoot/reference) to consider those fields?

**Review by timuthy — CHANGES_REQUESTED** [review]

File: pkg/admissioncontroller/webhook/admission/authorizationconfig/handler.go (hunk 6 above)
Comment (rfranzke) [reply pkg/admissioncontroller/webhook/admission/authorizationconfig/handler.go:97]: I don't remember exactly why this was introduced, but it's the same in the other admission handlers for audit policy and authn config. Let's keep it for now.

File: pkg/admissioncontroller/webhook/admission/authorizationconfig/handler.go (hunk 10 above)
Comment (rfranzke) [reply pkg/admissioncontroller/webhook/admission/authorizationconfig/handler.go:229]: > Where is the kubeconfig secret checked, i.e. the secret exists 

Good question, I don't think we have such check for several fields, see https://github.com/gardener/gardener/blob/24f0bea3fdfa8def827b882f8a2147ccdb384a07/plugin/pkg/global/resourcereferencemanager/admission.go#L883-L974 I've now added it.

> + `data["kubeconfig"]` is available?

This is covered by the generic https://github.com/gardener/gardener/tree/master/pkg/admissioncontroller/webhook/admission/kubeconfigsecret handler.

**Review by timuthy — COMMENTED** [review]
Love it, thank you 🚀 
/lgtm

File: docs/usage/shoot/shoot_access.md [hunk 12]
```diff
@@ -182,7 +182,67 @@ data:
...
+    kubeAPIServer:
+      structuredAuthorization:
+        configMapName: name-of-configmap-containing-authorization-config
+        kubeconfigs:
+        - authorizerName: my-webhook
+          secretName: webhook-kubeconfig
+```
+
+The `configMapName` references a user created `ConfigMap` in the project namespace containing the `AuthorizationConfiguration` in it's `config.yaml` data field.
+Here is an example of such `ConfigMap`:
```
Comment (oliver-goetz) [inline docs/usage/shoot/shoot_access.md:206]: The following code snippet does not include the ConfiMap only, but also a secret. This is a bit confusing.

File: pkg/admissioncontroller/webhook/admission/authorizationconfig/handler.go [hunk 13]
```diff
@@ -0,0 +1,250 @@
...
+	if err != nil {
+		return admission.Errored(http.StatusUnprocessableEntity, err)
+	}
+
+	if err = h.getOldObject(request, oldConfigMap); err != nil {
+		return admission.Errored(http.StatusInternalServerError, err)
+	}
+	if apiequality.Semantic.DeepEqual(configMap.Data, oldConfigMap.Data) {
+		return admissionwebhook.Allowed("authorization configuration or kubeconfig secret names not changed")
+	}
```
Comment (ScheererJ) [inline pkg/admissioncontroller/webhook/admission/authorizationconfig/handler.go]: Why mention the `kubeconfig` secret in this context when checking only the config map?

File: pkg/admissioncontroller/webhook/admission/authorizationconfig/handler.go [hunk 14]
```diff
@@ -0,0 +1,250 @@
...
+}
+
+func validateAuthorizationConfigurationSemantics(authorizationConfiguration string, kubeconfigReferences []gardencore.AuthorizerKubeconfigReference, verifyKubeconfigReferences bool) (errCode int32, err error) {
+	authConfigObj, schemaVersion, err := configDecoder.Decode([]byte(authorizationConfiguration), nil, nil)
+	if err != nil {
+		return http.StatusUnprocessableEntity, fmt.Errorf("failed to decode the provided authorization configuration: %w", err)
+	}
+	authConfig, ok := authConfigObj.(*apiserver.AuthorizationConfiguration)
+	if !ok {
+		return http.StatusInternalServerError, fmt.Errorf("failure to cast to authorization configuration type: %v", schemaVersion)
```
Comment (ScheererJ) [inline pkg/admissioncontroller/webhook/admission/authorizationconfig/handler.go]:
```suggestion
		return http.StatusInternalServerError, fmt.Errorf("failed to cast to authorization configuration type: %v", schemaVersion)
```

File: pkg/component/kubernetes/apiserver/authorization.go [hunk 15]
```diff
@@ -0,0 +1,135 @@
...
+			Name: volumeNameStructuredAuthorizationConfig,
+			VolumeSource: corev1.VolumeSource{
+				ConfigMap: &corev1.ConfigMapVolumeSource{LocalObjectReference: corev1.LocalObjectReference{
+					Name: configMapAuthorizationConfig.Name,
+				}},
+			},
+		})
+	} else {
+		// TODO: Delete this else-branch and everything related to it once we only support a Kubernetes version which
+		//  promotes the StructuredAuthorizationConfiguration feature gate to GA (1.32+ or higher).
```
Comment (ScheererJ) [inline pkg/component/kubernetes/apiserver/authorization.go]: Is the leading space in the second line intentional. It looks a bit strange.

**Review by ScheererJ — COMMENTED** [review]
Thanks a lot for providing this useful extension. I only found a few minor nits.

File: pkg/component/kubernetes/apiserver/authorization.go (hunk 15 above)
Comment (rfranzke) [reply pkg/component/kubernetes/apiserver/authorization.go]: Without this, the IDE does not proper color it:

image
vs.
image

File: docs/usage/shoot/shoot_access.md [hunk 16]
```diff
@@ -182,7 +182,67 @@ data:
...
+apiVersion: v1
+kind: Secret
+metadata:
+  name: webhook-kubeconfig
+  namespace: garden-my-project
+data:
+  kubeconfig: <base64-encoded-kubeconfig-for-authz-webhook>
+```
+
+The user is responsible for the validity of the configured authorizers.
```
Comment (oliver-goetz) [inline docs/usage/shoot/shoot_access.md:243]: Configuration changes in the configmap/secret apply after the next shoot reconciliation only, don't they? It would be nice to mention this.
The same might apply for Structured Authentication too.

**Review by oliver-goetz — COMMENTED** [review]
Nice PR, thanks 🚀 
I have remarks about the docs only.

File: pkg/operator/controller/garden/garden/components.go [hunk 17]
```diff
@@ -618,10 +619,13 @@ func (r *Reconciler) newKubeAPIServer(
...
 				Timeout:                                  metav1.Duration{Duration: 10 * time.Second},
-				FailurePolicy:                            apiserverv1beta1.FailurePolicyNoOpinion,
+				FailurePolicy:                            apiserverv1beta1.FailurePolicyDeny,
 				SubjectAccessReviewVersion:               "v1",
 				MatchConditionSubjectAccessReviewVersion: "v1",
-				MatchConditions:                          nil,
+				MatchConditions: []apiserverv1beta1.WebhookMatchCondition{{
+					// only intercept request from gardenlets and service accounts from seed namespaces
+					Expression: fmt.Sprintf("'%s' in request.groups || request.groups.exists(e, e.startsWith('%s%s'))", v1beta1constants.SeedsGroup, serviceaccount.ServiceAccountGroupPrefix, gardenerutils.SeedNamespaceNamePrefix),
+				}},
```
Comment (oliver-goetz) [inline pkg/operator/controller/garden/garden/components.go:631]: With this change we could try to get rid of deploying the virtual-kubeapiserver twice. There should be no calls to these namespaces until the Garden was reconciled for the first time.
I don't know if it works and it could be verified and implemented in an own PR.

Comment (rfranzke) [reply pkg/operator/controller/garden/garden/components.go:631]: That's a good point. I added a TODO statement. I think we can only do it once we only support Kubernetes versions of 1.32+.
